alphabet/alphabet_contract.go

```diff
@@ -60,14 +78,143 @@ func _deploy(data interface{}, isUpdate bool) {
 	storage.Put(ctx, indexKey, args.index)
 	storage.Put(ctx, totalKey, args.total)
 
-	// initialize the way to collect signatures
-	storage.Put(ctx, notaryDisabledKey, args.notaryDisabled)
-	if args.notaryDisabled {
-		common.InitVote(ctx)
-		runtime.Log(args.name + " notary disabled")
+	runtime.Log(args.name + " contract initialized")
+}
+
+// re-initializes contract from non-notary to notary mode. Does nothing if
+// action has already been done. The function is called on contract update with
+// storage.Context and parameters from _deploy.
+//
+// switchToNotary panics if address of the Proxy contract (3rd parameter) is
+// missing or invalid. Otherwise, on success, the address is stored by
+// 'proxyScriptHash' key.
+//
+// If contract stores non-empty value by 'ballots' key, switchToNotary panics.
+// Otherwise, existing value is removed.
+//
+// switchToNotary removes value stored by 'notary' key.
+//
+// nolint:unused
+func switchToNotary(ctx storage.Context, args []interface{}) {
+	const notaryDisabledKey = "notary" // non-notary legacy
+	contractName := args[3].(string)
+
+	notaryVal := storage.Get(ctx, notaryDisabledKey)
+	if notaryVal == nil {
+		runtime.Log(contractName + " contract is already notarized")
+		return
+	} else if notaryVal.(bool) {
+		proxyContract := args[2].(interop.Hash160)
+		if len(proxyContract) != interop.Hash160Len {
+			panic("address of the Proxy contract is missing or invalid")
+		}
+
+		// get NNS contract (it always has ID=1 in the NeoFS Sidechain)
+		nnsContract := management.GetContractByID(1)
+		if nnsContract == nil {
+			panic("missing NNS contract")
+		}
+
+		resResolve := contract.Call(nnsContract.Hash, "resolve", contract.ReadOnly,
+			"proxy.neofs", 16, // TXT
+		)
+
+		records := resResolve.([]string)
+		if len(records) == 0 {
+			panic("did not find a record of the Proxy contract in the NNS")
+		} else if records[0] != address.FromHash160(proxyContract) {
+			const hexTable = "0123456789abcdef"
+			j := 0
+			for _, v := range proxyContract {
+				if records[0][j] != hexTable[v>>4] || records[0][j+1] != hexTable[v&0x0f] {
+					panic("wrong address of the Proxy contract" + records[0])
+				}
+				j += 2
+			}
+		}
+
+		if !common.TryPurgeVotes(ctx) {
+			panic("pending vote detected")
+		}
+
+		// distribute 75% of available GAS:
+		//  - 50% to Proxy contract
+		//  - the rest is evenly distributed between Inner Ring and storage nodes
```

Comment:
IMO this is the best place for it since everything below is related to GAS distribution, for example, we need netmap contract to calculated GAS amount only